src/core/Subscription.ts

```diff
@@ -0,0 +1,126 @@
+import {
+	AudioPlayer,
+	AudioPlayerStatus,
+	AudioResource,
+	createAudioPlayer,
+	entersState,
+	VoiceConnection,
+	VoiceConnectionDisconnectReason,
+	VoiceConnectionStatus,
+} from '@discordjs/voice';
+
+import { Track } from './track';
+
+function wait(time: number) {
+	return new Promise((resolve) => setTimeout(resolve, time).unref());
+}
+
+export class Subscription {
+	public readonly voiceConnection: VoiceConnection;
+
+	public readonly audioPlayer: AudioPlayer;
+
+	public queue: Track[];
+
+	public queueLock = false;
+
+	public readyLock = false;
+
+	public constructor(voiceConnection: VoiceConnection) {
+		this.voiceConnection = voiceConnection;
+		this.audioPlayer = createAudioPlayer();
+		this.queue = [];
+
+		this.voiceConnection.on('stateChange', async (_, newState) => {
+			if (newState.status === VoiceConnectionStatus.Disconnected) {
+				if (newState.reason === VoiceConnectionDisconnectReason.WebSocketClose && newState.closeCode === 4014) {
+					/*
+						If the WebSocket closed with a 4014 code, this means that we should not manually attempt to reconnect,
+						but there is a chance the connection will recover itself if the reason of the disconnect was due to
+						switching voice channels. This is also the same code for the bot being kicked from the voice channel,
+						so we allow 5 seconds to figure out which scenario it is. If the bot has been kicked, we should destroy
+						the voice connection.
+					*/
+					try {
+						await entersState(this.voiceConnection, VoiceConnectionStatus.Connecting, 5_000);
+					} catch {
+						this.voiceConnection.destroy();
+					}
+				} else if (this.voiceConnection.rejoinAttempts < 5) {
+					await wait(this.voiceConnection.rejoinAttempts++ * 5_000);
+					this.voiceConnection.rejoin();
+				} else {
+					this.voiceConnection.destroy();
+				}
+			} else if (newState.status === VoiceConnectionStatus.Destroyed) {
+				this.stop();
+			} else if (
+				!this.readyLock &&
+				(newState.status === VoiceConnectionStatus.Connecting || newState.status === VoiceConnectionStatus.Signalling)
+			) {
+				/*
+					In the Signalling or Connecting states, we set a 20 second time limit for the connection to become ready
+					before destroying the voice connection. This stops the voice connection permanently existing in one of these
+					states.
+				*/
+				this.readyLock = true;
+				try {
+					await entersState(this.voiceConnection, VoiceConnectionStatus.Ready, 20_000);
+				} catch {
+					if (this.voiceConnection.state.status !== VoiceConnectionStatus.Destroyed) {
+						this.voiceConnection.destroy();
+					}
+				} finally {
+					this.readyLock = false;
+				}
```

Comment:
We need to remember to change this at some point, because at the moment it prevents a 5 rejoin attempt limit and replaces it with a 20 second hard cap.